quantum/action_util.c

```diff
@@ -276,7 +282,7 @@ void send_keyboard_report(void) {
     static report_keyboard_t last_report;
 
     /* Only send the report if there are changes to propagate to the host. */
```

Comment:
Why is this logic in `send_keyboard_report()` in the first place?

Reply:
The internal state machine of QMK produces duplicated, mostly empty, reports which is not trivial to change. This doesn't happen for regular key presses but more advanced features like oneshot keys and mod-taps. You can have a look at the o.g. implementation and especially the tests for more details. https://github.com/qmk/qmk_firmware/pull/14065

Reply:
I can't find any instances of `send_keyboard_report()` being called without the report being modified immediately beforehand.

Reply:
If you disable the logic and run all unit-tests you'll observe lots of failing unit-tests with unsatisfied mock calls for empty reports e.g. this run of a tap dance which has multiple consecutive empty reports in a row:

```
ayers/TapDanceLayers.SingleTapFast/LayerMoveCustomFast failed!
[ LEVEL    ] [TIME] [EVENT]
[ INFO     ] 0      tapping term is 200ms
[ TRACE    ] 0      pressed:  TD(0)
[ TRACE    ] 0      1 keyboard task loop
[ TRACE    ] 1      released: TD(0) was pressed for 1ms
[ TRACE    ] 1      1 keyboard task loop
[ TRACE    ] 2      pressed:  SAFE_RANGE
[ TRACE    ] 2      1 keyboard task loop
[ TRACE    ] 2      report:   empty
[ TRACE    ] 2      report:   (KC_APPLICATION) []
[ TRACE    ] 2      report:   empty
[ TRACE    ] 2      report:   empty
[ TRACE    ] 2      report:   (KC_A) []
[ TRACE    ] 2      report:   empty
[ TRACE    ] 3      released: SAFE_RANGE was pressed for 1ms
[ TRACE    ] 3      1 keyboard task loop
[ TRACE    ] 3      report:   (KC_B) []
[ TRACE    ] 3      report:   empty
[ INFO     ] 4      test fixture clean-up start.
[ TRACE    ] 4      report:   empty
[ TRACE    ] 4      2000 keyboard task loops
[ TRACE    ] 2004   2000 keyboard task loops
[ INFO     ] 4004   test fixture clean-up end.
```

Reply:
or this quick tap unit test where a left shift and empty report is send twice:

```
QuickTap.tap_regular_key_while_mod_tap_key_is_held failed!
[ LEVEL    ] [TIME] [EVENT]
[ INFO     ] 4122   test fixture setup-up start.
[ INFO     ] 0      test fixture setup-up end.
[ INFO     ] 0      tapping term is 200ms
[ TRACE    ] 0      pressed:  MT(MOD_LSFT, KC_P)
[ TRACE    ] 0      1 keyboard task loop
[ TRACE    ] 1      pressed:  KC_A
[ TRACE    ] 1      1 keyboard task loop
[ TRACE    ] 2      released: KC_A was pressed for 1ms
[ TRACE    ] 2      1 keyboard task loop
[ TRACE    ] 3      released: MT(MOD_LSFT, KC_P) was pressed for 3ms
[ TRACE    ] 3      1 keyboard task loop
[ TRACE    ] 3      report:   () [KC_LEFT_SHIFT]
[ TRACE    ] 4      197 keyboard task loops
[ TRACE    ] 200    report:   () [KC_LEFT_SHIFT]
[ TRACE    ] 200    report:   (KC_A) [KC_LEFT_SHIFT]
[ TRACE    ] 200    report:   () [KC_LEFT_SHIFT]
[ TRACE    ] 200    report:   empty
[ INFO     ] 201    test fixture clean-up start.
[ TRACE    ] 201    report:   empty
[ TRACE    ] 201    2000 keyboard task loops
[ TRACE    ] 2201   2000 keyboard task loops
[ INFO     ] 4201   test fixture clean-up end
```

Reply:
This seems to me indicative of a larger issue which needs fixing, rather than adding these bandaids.

Reply:
The implementation of QMK behaviors that act upon the current state of the keyboard (e.g. tap dances) are way easier to implement if they don't have to care about the previous state of the resulting HID report. For instance the tap dance callback function that is called on timeout would be much more involved if it had to determine if it would really have to send a report or not:

```c
static inline void process_tap_dance_action_on_reset(tap_dance_action_t *action) {
    _process_tap_dance_action_fn(&action->state, action->user_data, action->fn.on_reset);
    del_weak_mods(action->state.weak_mods);
#ifndef NO_ACTION_ONESHOT
    del_mods(action->state.oneshot_mods);
#endif
    send_keyboard_report();
    action->state = (const tap_dance_state_t){0};
}
```

I see the current de-duplication in `send_keyboard_report()` more as a `reduce` operation at the end of interacting with the internal state machine just prior to output. There are certainly cases where `send_keyboard_report()` calls are real implementation bugs, but most of the time the callers benefit from this de-duplication.

My proposal would be to rename `send_keyboard_report_forced()` to `send_keyboard_report()` and come up with a different name for the "current" `send_keyboard_report()` that reflects the de-duplication behavior.

Reply:
For that example at least, it can be slightly tweaked into:
```c
static inline void process_tap_dance_action_on_reset(tap_dance_action_t *action) {
    _process_tap_dance_action_fn(&action->state, action->user_data, action->fn.on_reset);

#ifndef NO_ACTION_ONESHOT
    del_weak_mods(action->state.weak_mods);
    unregister_mods(action->state.oneshot_mods);
#else
    unregister_weak_mods(action->state.weak_mods);
#endif

    action->state = (const tap_dance_state_t){0};
}
```
Which is perhaps still not ideal, but it's better than keeping track of previous state like you said.

Reply:
The question remains though if you are convinced by my argument and subsequent proposal to keep the de-duplication and adjust the naming?

Reply:
I'm resolving this conversation as the root cause of the issue something completely different.